keps/sig-autoscaling/20191111-autoscaling-with-multiple-thresholds.md

```diff
@@ -0,0 +1,202 @@
+<img width="911" alt="Upscale Capping" src="https://user-images.githubusercontent.com/7433560/63385168-f46c7900-c38f-11e9-9e7c-1a7796afd31e.png">
+
+In this similar example, we avoid downscaling by too many replicas, and can use the same set of metrics to guarantee that we only scale down by a reasonable amount of replicas.
+<img width="911" alt="Downscale Capping" src="https://user-images.githubusercontent.com/7433560/63385340-44e3d680-c390-11e9-91d4-35b2f8a912ad.png">
+
+The last options available to restrict scaling are `downscaleForbiddenWindowSeconds` and `upscaleForbiddenWindowSeconds` in seconds that represent respectively how much time to wait before scaling down and scaling up after **any scaling event**. We do not compare the `upscaleForbiddenWindowSeconds` to the last time we only upscaled, so a downscale event will prevent an upscale event for a duration of `upscaleForbiddenWindowSeconds`.
+
+```
+    downscaleForbiddenWindowSeconds: 60
+    upscaleForbiddenWindowSeconds: 30
+```
+
+In the following example we can see that the recommended number of replicas is ignored if we are in a cooldown period. The downscale cooldown period can be visualised with `watermarkpodautoscaler.wpa_controller_transition_countdown{transition:downscale}`, and is represented in yellow on the graph below. We can see that it is significantly higher than the upscale cooldown period (`transition:upscale`) in dark orange on our graph. As soon as we are recommended to scale, only if the appropriate cooldown window is over, will we scale. This will reset both countdowns.
+<img width="911" alt="Forbidden Windows" src="https://user-images.githubusercontent.com/7433560/63389864-a14cf300-c39c-11e9-9ad5-8308af5442ad.png">
+
+### Implementation Details/Notes/Constraints
+
+First and foremost, upon making a measurement (i.e. querying the Metrics Server or Custom/External Metrics Server) we compare the value to both thresholds, we can only suggest upscaling if the value is at least higher than the high threshold and a downscale event if it is at least lower than the low threshold.
+
+As we add a second threshold, it seems confusing and error prone to keep the logic of the `targetAverageValue` and the `targetValue` nomenclature for those two thresholds. e.g. have the LowThreshold/LowAverageThreshold and deal with cases where a user could set a HighThreshold and a LowAverageThreshold. 
+So we decided to implement a separate option in the spec, `algorithm`.
+
+Depending on your use case, you might want to consider one or the other. The naming of the algorithm is from the point of view of the controller.
```

Comment:
Providing named algorithms as part of the API adds too much coupling to the implementation.  I would prefer to maintain the API as a description of the desired outcome (metric targets) and some constraints (behavior structures).

Reply:
As I am going to suggest sticking to the threshold logic, we won't need the algorithm names anymore.